### PR TITLE
Instant Region Change

### PR DESCRIPTION
**Pull request by bagusnl**
- Instantly change the region when selected game or region is changed.
- Hide the "Change" button.
- Enabled by default when "Show Region Warning" is disabled
 
image

# Video Showcase

https://github.com/CollapseLauncher/Collapse/assets/28079733/05ede866-d9e7-413e-81ae-f327ad35ccb8

### CODE REVIEW

**Review by bagusnl — COMMENTED** [review]
All good on my end, requesting review from anyone on @CollapseLauncher/core-maintainers

**Review by shatyuka — COMMENTED** [review]
Unable to change region if we turn off "instant change region" toggle unless a restart. Maybe we can preserve the old behavior (instant change ON), while showing "instant change OFF" in the settings page.

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs [hunk 1]
```diff
@@ -78,9 +78,12 @@ public SettingsPage()
             if (IsAppLangNeedRestart)
                 AppLangSelectionWarning.Visibility = Visibility.Visible;
 
-            if (IsChangeRegionWarningNeedRestart)
+            if (IsChangeRegionWarningNeedRestart)
```
Comment (shatyuka) [inline CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:81]: unexpected space here

File: Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs [hunk 2]
```diff
@@ -266,6 +267,12 @@ public static bool IsShowRegionChangeWarning
             set => SetAndSaveConfigValue("ShowRegionChangeWarning", value);
         }
 
+        public static bool IsInstantRegionChange
+        {
+            get => GetAppConfigValue("UseInstantRegionChange").ToBool();
+            set => SetAndSaveConfigValue("UseInstantRegionChange", value);
+        }
+
```
Comment (shatyuka) [inline Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs:275]: check `IsInstantRegionNeedRestart` and current mode before we return the config value? need to save the config values at startup.

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs [hunk 3]
```diff
@@ -788,16 +791,42 @@ private bool IsIncludeGameLogs
...
+            get
+            { 
+                var value = LauncherConfig.IsShowRegionChangeWarning;
+
+                ToggleChangeRegionInstant.Visibility = !value ? Visibility.Visible : Visibility.Collapsed;
+                return value;
+            }
             set
             {
                 IsChangeRegionWarningNeedRestart = true;
```
Comment (shatyuka) [inline CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:803]: hide the warning if we switch it back

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs [hunk 4]
```diff
@@ -788,16 +791,42 @@ private bool IsIncludeGameLogs
...
+                }
             }
         }
 
+        private bool IsInstantRegionChange
+        {
+            get => LauncherConfig.IsInstantRegionChange;
+            set
+            {
+                IsInstantRegionNeedRestart            = true;
```
Comment (shatyuka) [inline CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:824]: hide the warning if we switch it back

File: Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs (hunk 2 above)
Comment (bagusnl) [reply Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs:275]: I'm not sure if that's necessary

For the comment body about can't change region when feature is disabled before restart, i fixed it with this 

```csharp
        private static bool? _cachedIsInstantRegionChange = null;
        public static bool IsInstantRegionChange
        {
            get
            {
                _cachedIsInstantRegionChange ??= GetAppConfigValue("UseInstantRegionChange").ToBool();
                return (bool)_cachedIsInstantRegionChange;
            }
            set => SetAndSaveConfigValue("UseInstantRegionChange", value);
        }
```

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs (hunk 1 above)
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/449/commits/dd2901c974e260044f78024631e13d69988b9e12

File: Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs (hunk 2 above)
Comment (bagusnl) [reply Hi3Helper.Core/Classes/Shared/Region/LauncherConfig.cs:280]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/449/commits/dd2901c974e260044f78024631e13d69988b9e12

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs (hunk 3 above)
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/449/commits/dd2901c974e260044f78024631e13d69988b9e12

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs (hunk 4 above)
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/449/commits/dd2901c974e260044f78024631e13d69988b9e12

**Review by shatyuka — APPROVED** [review]

**Pull request merged**